File: backend/app/auth/password_policy.py
import re


PASSWORD_MIN_LENGTH = 8


def get_password_policy_violations(
    password: str,
) -> list[str]:
    violations: list[str] = []

    if len(password) < PASSWORD_MIN_LENGTH:
        violations.append(
            f"at least {PASSWORD_MIN_LENGTH} characters"
        )

    if not re.search(r"[A-Z]", password):
        violations.append(
            "one uppercase letter"
        )

    if not re.search(r"[a-z]", password):
        violations.append(
            "one lowercase letter"
        )

    if not re.search(r"[0-9]", password):
        violations.append(
            "one digit"
        )

    if not re.search(r"[^A-Za-z0-9]", password):
        violations.append(
            "one special character"
        )

    return violations

File: frontend/src/utils/passwordPolicy.ts
export type PasswordPolicyCheck = {
  label: string;
  isValid: boolean;
};

const PASSWORD_MIN_LENGTH = 8;

export function getPasswordPolicyChecks(
  password: string,
): PasswordPolicyCheck[] {
  return [
    {
      label: `At least ${PASSWORD_MIN_LENGTH} characters`,
      isValid: password.length >= PASSWORD_MIN_LENGTH,
    },
    {
      label: "One uppercase letter",
      isValid: /[A-Z]/.test(password),
    },
    {
      label: "One lowercase letter",
      isValid: /[a-z]/.test(password),
    },
    {
      label: "One digit",
      isValid: /[0-9]/.test(password),
    },
    {
      label: "One special character",
      isValid: /[^A-Za-z0-9]/.test(password),
    },
  ];
}

export function isPasswordPolicyValid(
  password: string,
): boolean {
  return getPasswordPolicyChecks(password).every(
    (check) => check.isValid,
  );
}